runtime/src/parachains.rs

```diff
@@ -726,26 +725,27 @@ impl<T: Trait> Module<T> {
 				"Not enough validity attestations"
 			);
 
+			ensure!(
+				candidate.validity_votes.len() <= authorities.len(),
+				"The number of attestations exceeds the number of authorities"
+			);
+
 			let fees = candidate.candidate().fees;
 			T::ParachainCurrency::deduct(para_id, fees)?;
 
 			let mut candidate_hash = None;
 			let mut encoded_implicit = None;
 			let mut encoded_explicit = None;
 
-			// track which voters have voted already, 1 bit per authority.
-			let mut track_voters = bitvec![0; authorities.len()];
-			for (auth_index, validity_attestation) in &candidate.validity_votes {
-				let auth_index = *auth_index as usize;
-				// protect against double-votes.
-				match validator_group.iter().find(|&(idx, _)| *idx == auth_index) {
-					None => return Err("Attesting validator not on this chain's validation duty."),
-					Some(&(idx, _)) => {
-						if track_voters.get(idx) {
-							return Err("Voter already attested validity once")
-						}
-						track_voters.set(idx, true)
-					}
+			for ((auth_index, _), validity_attestation) in candidate.validator_indices
+				.iter()
+				.enumerate()
+				.filter(|(_, bit)| *bit)
+				.zip(candidate.validity_votes.iter())
```

Comment:
i don't think this is right - this allows `candidate.validity_votes` to contain an arbitrary amount of elements beyond the number of set bits.

You could add another `enumerate` so you have `(vote_index, (auth_index, _))` and use `candidate.validity_votes.get(vote_index)` within the body of the loop, and if it's `None`, bail.